src/main/scala/cognite/spark/v1/AssetHierarchyBuilder.scala

```diff
@@ -174,86 +124,55 @@ class AssetHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
+                .map(missing => roots.filter(_.parentExternalId == missing).minBy(_.externalId))
+                .map(_.externalId)
+            InvalidNodeReferenceException(missingNodes, referencingNodes)
+        }),
+    ).void
 
   def buildSubtrees(trees: Vector[AssetSubtree]): IO[Unit] =
     for {
       // fetch existing roots and update or insert them first
-      cdfRoots <- fetchCdfRoots(trees.map(_.root.externalId))
+      cdfRoots <- fetchCdfAssets(trees.map(_.root.externalId))
+      // fetch existing nodes in CDF in order to know if we need to insert or update them
+      cdfAssets <- fetchCdfAssets(trees.flatMap(_.nodes).map(_.externalId))
+      (toInsert, toUpdate, _) = nodesToInsertUpdate(trees.flatMap(_.nodes), cdfAssets)
+
       upsertedRoots <- upsertRoots(trees.map(_.root), cdfRoots)
 
-      // then fetch all the subtrees to know if we should insert, update or delete them
-      // TODO: merge the subtree requests
-      cdfTrees <- trees.parTraverse(t => fetchCdfSubtree(t.root))
-      (toDelete, toInsert, toUpdate) = trees
-        .zip(cdfTrees)
-        .map((nodesToDeleteInsertUpdate _).tupled)
-        .unzip3
-
-      _ <- (
-        delete(toDelete.flatten, deleteMissingAssets),
-        insert(toInsert.flatten, trees, upsertedRoots),
-        update(toUpdate.flatten)
-      ).parMapN((_, _, _) => ())
+      _ = trees.foreach(validateNoRootChange(_, upsertedRoots, cdfAssets))
+
+      _ <- deleteMissingChildren(trees, deleteMissingAssets)
+      // insert must be done before update, since the updated node can reference node that is to be created
+      _ <- insert(toInsert)
+      _ <- update(toUpdate)
     } yield ()
 
-  def insert(
-      toInsert: Seq[AssetsIngestSchema],
-      trees: Vector[AssetSubtree],
-      createdRoots: Vector[Asset]): IO[Unit] = {
+  def insert(toInsert: Seq[AssetsIngestSchema]): IO[Vector[Asset]] = {
     val assetCreatesToInsert = toInsert.map(toAssetCreate)
     // Traverse batches in order to ensure writing parents first
     assetCreatesToInsert
       .grouped(batchSize)
-      .toList
+      .toVector
       .traverse(
         client.assets
           .create(_)
-          .flatMap(x => incMetrics(itemsCreated, x.size))
+          .flatTap(x => incMetrics(itemsCreated, x.size))
       )
-      .void
-      .adaptErr {
-        case ex: CdpApiException if ex.duplicated.isDefined =>
-          val ids =
-            ex.duplicated.get
-              .flatMap(decodeCogniteId)
-              .collect { case CogniteExternalId(id) => id }
-          // If we've hit this case, the assets already exist in CDF, but under a different root asset.
-          // Since we didn't find them under the subtree, we assumed they did not exist, and attempted to to insert them,
-          // resulting in this error.
-          if (ids.nonEmpty) {
-            InvalidRootChangeException.createFromTreeList(ex, ids, trees, createdRoots)
```

Comment:
We now validate for this before even doing the insert, so no need for error-parsing logic